src/server/snapshot.cc

```diff
@@ -61,6 +61,47 @@ void SliceSnapshot::Start(bool stream_journal, const Cancellation* cll) {
   });
 }
 
+void SliceSnapshot::StartIncremental(Context* cntx, LSN start_lsn) {
+  auto* journal = db_slice_->shard_owner()->journal();
+  DCHECK(journal);
+
+  serializer_ = std::make_unique<RdbSerializer>(compression_mode_);
+
+  snapshot_fb_ =
+      fb2::Fiber("incremental_snapshot", [this, journal, cntx, lsn = start_lsn]() mutable {
+        // The LSN we get from the replica is the last entry that
+        // was processed, so no need to send it.
+        lsn++;
+        while (!cntx->IsCancelled() && journal->IsLSNInBuffer(lsn)) {
+          // TODO: Should we send multiple entries together?
+          serializer_->WriteJournalEntry(journal->GetEntry(lsn));
+          PushSerializedToChannel(false);
+          lsn++;
+        }
+        // This check is safe, but it is not trivially safe.
+        // We rely here on the fact that JournalSlice::AddLogRecord can
+        // only preempt while holding the callback lock.
+        // That guarantees that if we have processed the last LSN the callback
+        // will only be added after JournalSlice::AddLogRecord has finished
+        // iterating its callbacks and we won't process the record twice.
+        // We have to make sure we don't preempt ourselves before registering the callback!
+        if (journal->GetLsn() == (lsn - 1)) {
+          {
+            FiberAtomicGuard fg;
+            serializer_->SendFullSyncCut();
+          }
+          auto journal_cb = absl::bind_front(&SliceSnapshot::OnJournalEntry, this);
+          journal_cb_id_ = journal->RegisterOnChange(std::move(journal_cb));
+          PushSerializedToChannel(true);
+        } else {
+          // We stopped but we didn't manage to send the whole stream.
+          cntx->ReportError(std::make_error_code(errc::state_not_recoverable),
+                            "Partial sync was not successful");
+          Cancel();
```

Comment:
Same here, the error report will trigger the cleanup handler which in turn will call cancel on all snapshots, so in total it will be called twice. But I remember I guarded Cancel against multiple calls because I had such bugs already 😄